File: src/ros2_bridge/mappers/foxglove_msgs/raw_image.rs
//! Typed mapper for `foxglove_msgs/msg/RawImage`.

use crate::ros2_bridge::mapper::TypedTopicMapper;

pub(crate) fn raw_image_to_bus(msg: ros_env::foxglove_msgs::msg::RawImage) -> crate::foxglove_msgs::msg::v1::RawImage {
    crate::foxglove_msgs::msg::v1::RawImage {
        timestamp: Some(crate::ros2_bridge::mappers::convert::time_to_timestamp(msg.timestamp)),
        frame_id: crate::ros2_bridge::mappers::convert::from_ros_string(msg.frame_id),
        width: msg.width,
        height: msg.height,
        encoding: crate::ros2_bridge::mappers::convert::from_ros_string(msg.encoding),
        step: msg.step,
        data: crate::ros2_bridge::mappers::convert::IntoU8Vec::into_u8_vec(msg.data),
    }
}

pub(crate) fn raw_image_to_ros(bus: crate::foxglove_msgs::msg::v1::RawImage) -> ros_env::foxglove_msgs::msg::RawImage {
    ros_env::foxglove_msgs::msg::RawImage {
        timestamp: crate::ros2_bridge::mappers::convert::timestamp_to_time(bus.timestamp.unwrap_or_default()),
        frame_id: crate::ros2_bridge::mappers::convert::to_ros_string(bus.frame_id),
        width: bus.width,
        height: bus.height,
        encoding: crate::ros2_bridge::mappers::convert::to_ros_string(bus.encoding),
        step: bus.step,
        data: crate::ros2_bridge::mappers::convert::FromByteSeq::from_byte_seq(bus.data),
    }
}

#[derive(Clone, Copy, Debug, Default)]
pub struct FoxgloveMsgsRawImageMapper;

impl TypedTopicMapper for FoxgloveMsgsRawImageMapper {
    type Ros = ros_env::foxglove_msgs::msg::RawImage;
    type Bus = crate::foxglove_msgs::msg::v1::RawImage;

    fn type_name(&self) -> &'static str {
        "foxglove_msgs/msg/RawImage"
    }

    fn ros_to_bus(&self, msg: Self::Ros) -> crate::errors::Result<Self::Bus> {
        Ok(raw_image_to_bus(msg))
    }

    fn bus_to_ros(&self, msg: Self::Bus) -> crate::errors::Result<Self::Ros> {
        Ok(raw_image_to_ros(msg))
    }
}

File: bindings/python/robot_bus/ros2_bridge/mappers/foxglove_msgs/raw_image.py
"""Generated mapper for `foxglove_msgs/msg/RawImage`."""

from __future__ import annotations

from robot_bus.ros2_bridge.mappers import _convert

_BusMsg = None


def _bus_cls():
    global _BusMsg
    if _BusMsg is None:
        from robot_bus.foxglove_msgs.msg.v1 import RawImage as BusMsg

        _BusMsg = BusMsg
    return _BusMsg


def raw_image_to_bus(msg):
    BusMsg = _bus_cls()
    bus = BusMsg()
    bus.timestamp = _convert.time_to_timestamp(msg.timestamp)
    bus.frame_id = str(msg.frame_id)
    bus.width = msg.width
    bus.height = msg.height
    bus.encoding = str(msg.encoding)
    bus.step = msg.step
    bus.data = bytes(msg.data)
    return bus


def raw_image_to_ros(bus):
    from foxglove_msgs.msg import RawImage as RosMsg

    out = RosMsg()
    out.timestamp = _convert.timestamp_to_time(bus.timestamp)
    out.frame_id = str(bus.frame_id)
    out.width = bus.width
    out.height = bus.height
    out.encoding = str(bus.encoding)
    out.step = bus.step
    out.data = bytes(bus.data)
    return out


class FoxgloveMsgsRawImageMapper:
    _ros_type = None

    def ros_msg_type(self):
        cls = type(self)
        if cls._ros_type is None:
            from foxglove_msgs.msg import RawImage as RosMsg

            cls._ros_type = RosMsg
        return cls._ros_type

    def ros_to_bus(self, msg) -> bytes:
        return raw_image_to_bus(msg).SerializeToString()

    def bus_to_ros(self, payload: bytes):
        BusMsg = _bus_cls()
        bus = BusMsg()
        bus.ParseFromString(payload)
        return raw_image_to_ros(bus)
